src/memory.cpp

```diff
@@ -12,17 +12,31 @@ static unsigned int ulog2(unsigned int numBlocks) {
   return (unsigned int) ceil(log2(std::max(numBlocks, (unsigned int) 2)));
 }
 
-Memory* Memory::create(unsigned int numBlocks, MemEncoding encoding) {
+static unsigned int norm2(unsigned int number) {
+  return (unsigned int) pow(2, ulog2(number));
+}
+
+Memory* Memory::create(
+    unsigned int numGlobalBlocks, unsigned int numLocalBlocks,
+    MemEncoding encoding) {
   switch(encoding) {
   case MemEncoding::SINGLE_ARRAY:
-    return new SingleArrayMemory(numBlocks);
+    return new SingleArrayMemory(norm2(numGlobalBlocks), norm2(numLocalBlocks));
   case MemEncoding::MULTIPLE_ARRAY:
-    return new MultipleArrayMemory(numBlocks);
+    return new MultipleArrayMemory(norm2(numGlobalBlocks), norm2(numLocalBlocks));
   default:
     llvm_unreachable("Unknown memory encoding");
   }
 }
 
+expr Memory::isGlobalBlock(const expr &bid) const {
```

Comment:
We can check whether a bid is non-local by masking MSB and comparing it to zero.
ult is expensive, maybe?

Reply:
This is because I assumed the number of global blocks and local blocks could be different. Is it too much assumption..?

Reply:
I agree they can be different, but is the local/nonlocal bid distinguished by MSB?
Or maybe we can try a different encoding which is suggested in this patch..? what do you think?

Reply:
Aha then separating `isLocal` bit from bid bits is a good idea! I'll update this :)